api/app/models/job_file.rb

```diff
@@ -111,18 +119,23 @@ def size(recalculate: false)
     @size ||= payload.length
   end
 
+  # Checks the file exists and the user has permission to access it
+  # NOTE: This is not cached to prevent it becoming stale
+  #       As it makes a FlightJobScriptAPI::SystemCommand, the cache
+  #       should be pre-warmed where ever possible before calling 'find!'
   def exists?
-    payload # Attempts to load the payload
-    true
-  rescue FlightJobScriptAPI::CommandError
-    # The file does not exists due to a bad name
-    return false unless payload_command
-
-    # The file or job legitimately does not exist
-    return false if [20, 23].include? payload_command.exitstatus
-
-    # Something else has gone wrong, re-raise the error
-    raise $!
+    return false unless path
+    return false unless File.exists?(path)
+
+    # Check the file permissions
+    # NOTE: The permission check is required to prevent user's polling
+    #       the filesystem via the API
+    cmd = FlightJobScriptAPI::SystemCommand.check_readable(id, path, user: @user)
+    return true if cmd.exitstatus == 0
+    return false if cmd.exitstatus == 20
+
+    # Something else went wrong
+    raise FlightJobScriptAPI::CommandError, "Unexpectedly failed to determine if the file exists: #{id}"
```

Comment:
It would crash any request in which the line is reached. However this shouldn't happen in practice.

I've added an additional existence check as the user: https://github.com/openflighthpc/flight-job-script-service/pull/38/commits/b07234147c34560a13a739d8f5161014c3070f18

The command should now only exit `0` or `20` unless something catastrophic happens.

---

The `exists?` command is now called in two places:

* `JobFile.find!` - Which already as a safe alternative method: `find`, and
* `JobFile#payload`

The `payload` is not serialised unless explicitly requested by the user, so a `GET /jobs` nor `GET /jobs/:id` will be effected by it.

Only when `fields[files]=payload,...` is specified will the `payload` method be called. This method already raises `CommandError` if the `file` doesn't exist. This would only happen if:

* A "non existent file" got cached for some reason, OR
* A race condition has occurred between the `cache` and an external `rm`

In either case, it is a server error and 50x is an appropriate response.